Fix rubocop offenses

KapilSachdev opened the pull request:
Helps move forward with #1345 
Cop fixes are broken into multiple commits to ease reviewing.

<details>
  <summary markdown="span">Checklist of Fixed Cops:</summary>

- [x] Bundler/OrderedGems
- [x] Layout/ArgumentAlignment
- [x] Layout/CaseIndentation
- [x] Layout/ClosingHeredocIndentation
- [x] Layout/ClosingParenthesisIndentation
- [x] Layout/DotPosition
- [x] Layout/ElseAlignment
- [x] Layout/EmptyLineAfterGuardClause
- [x] Layout/EmptyLineAfterMagicComment
- [x] Layout/EmptyLines
- [x] Layout/EmptyLinesAroundBlockBody
- [x] Layout/EmptyLinesAroundModuleBody
- [x] Layout/EndAlignment
- [x] Layout/HashAlignment
- [x] Layout/IndentationConsistency
- [x] Layout/IndentationWidth
- [x] Layout/LeadingCommentSpace
- [x] Layout/LineLength
- [x] Layout/MultilineBlockLayout
- [x] Layout/MultilineMethodCallBraceLayout
- [x] Layout/MultilineOperationIndentation
- [x] Layout/SpaceAfterComma
- [x] Layout/SpaceAroundEqualsInParameterDefault
- [x] Layout/SpaceAroundOperators
- [x] Layout/SpaceBeforeComment
- [x] Layout/SpaceInLambdaLiteral
- [x] Layout/SpaceInsideArrayLiteralBrackets
- [x] Layout/SpaceInsideHashLiteralBraces
- [x] Layout/SpaceInsidePercentLiteralDelimiters
- [x] Layout/SpaceInsideRangeLiteral
- [x] Layout/SpaceInsideReferenceBrackets
- [x] Layout/TrailingEmptyLines
- [x] Layout/TrailingWhitespace
- [x] Lint/AmbiguousBlockAssociation
- [x] Lint/MissingCopEnableDirective
- [x] Lint/NestedMethodDefinition
- [x] Lint/RedundantCopDisableDirective
- [x] Lint/RedundantRequireStatement
- [x] Lint/UnusedBlockArgument
- [x] Lint/UnusedMethodArgument
- [x] Metrics/ModuleLength
- [x] Naming/MemoizedInstanceVariableName
- [x] Naming/RescuedExceptionsVariableName
- [x] Rails/Output
- [x] Rails/Presence
- [x] Security/Eval
- [x] Security/Open
- [x] Style/ClassCheck
- [x] Style/CollectionMethods
- [x] Style/ConditionalAssignment
- [x] Style/EmptyMethod
- [x] Style/EvalWithLocation
- [x] Style/ExpandPathArguments
- [x] Style/FormatStringToken
- [x] Style/HashSyntax
- [x] Style/IfInsideElse
- [x] Style/InverseMethods
- [x] Style/MutableConstant
- [x] Style/ParallelAssignment
- [x] Style/RedundantBegin
- [x] Style/RedundantCondition
- [x] Style/RedundantInterpolation
- [x] Style/RedundantSelf
- [x] Style/RedundantSort
- [x] Style/RescueStandardError
- [x] Style/SafeNavigation
- [x] Style/StringLiterals
- [x] Style/StringLiteralsInInterpolation
- [x] Style/SymbolProc
- [x] Style/TrailingCommaInArguments (questionable)
- [x] Style/TrailingCommaInArrayLiteral
- [x] Style/TrailingCommaInHashLiteral

</details>

KapilSachdev commented:
@mcmire do you prefer multiple PR's or a single one for all?

KapilSachdev commented:
Pending:
| Offense Count | Cop |
|----|-------------------------------------|
| 9  | Metrics/AbcSize                                           |
| 8  | Metrics/CyclomaticComplexity                     |
| 7  | Metrics/MethodLength                                 |
| 4  | Metrics/BlockNesting                                   |
| 3  | Metrics/PerceivedComplexity                      |

utkarsh2102 reviewed (commented):
Hey,

Nice work!
Maybe just have:
```
Style/StringLiterals:
  Enabled: true
  EnforcedStyle: double_quotes

Style/StringLiteralsInInterpolation:
  Enabled: true
  EnforcedStyle: double_quotes
```

It's better to use `double_quotes` and also this codebase uses it anyway!

mcmire commented on spec/unit/shoulda/matchers/active_record/serialize_matcher_spec.rb:
```diff
@@ -79,8 +79,8 @@ def with_serialized_attr(type = nil)
     define_class(name) do
-      def load(*); end
-      def dump(*); end
+      def load(*); end # rubocop:disable Lint/NestedMethodDefinition
```
Interesting but okay 🤔

mcmire commented on lib/shoulda/matchers/active_model/qualifiers/ignoring_interference_by_writer.rb:
```diff
@@ -6,7 +6,7 @@ module Qualifiers
           attr_reader :ignore_interference_by_writer
 
-          def initialize(*args)
+          def initialize(*_args)
```
Since the name isn't important here, if there is a way to enforce this in Rubocop, can we just spell this is as `*`? If not, then no big deal — I don't want you to have to change all of these. Just wondering.

KapilSachdev replied:
I will change this.

mcmire reviewed (commented):
Did a pass-through of the whole PR. In hindsight perhaps it would have been better as a succession of PRs, but we've already come this far, so it's not a big deal. I know I left a lot of comments, but let me see if I can summarize them:

* Can we stick to an 80-character line length? I know that seems small in the days of large screens, but that's what was being followed before (and on a laptop it's still not so small). :) Also, can we re-enable line length in spec files? 
* When it comes to indentation, in general I (and to my knowledge this was a rule at thoughtbot, so it doesn't come from nowhere) try not to add more spaces than is necessary; instead the number of spaces should be an even number. What this means is that in cases where you have an assignment or are using an operator, and the right hand side needs to be spread across multiple lines, it's more maintainable, and an easier rule to remember IMO, to write:

``` ruby
foo <<
  if bar
    "baz"
  else
    "qux"
  end
```

rather than:

``` ruby
foo << if bar
         "baz"
       else
         "qux"
       end
```
* In a similar vein, there are some changes around indentation of arguments to methods that span multiple lines. It has also been an unofficial rule that if any argument to a method is on its own line, then all arguments should be on their own lines. Is there any way to enforce this in Rubocop? 
* There seem to be a lot of changes around trailing commas. This was a change I made to the Rubocop config a while back but which I now regret (it creates more of a pain point than it fixes). Considering there are a lot of files that now violate this rule, I'm wondering if that would be better to revert this cop (not sure how much trouble it is).
* There are a few cases where a long string was reformatted to fit across multiple lines but now contain line breaks that weren't there before, so make sure to change those back.
* There are a few false negatives that I ran across, so make sure to look out for those and change those back too.

mcmire commented:
To answer your questions about the pending violations, here's what I would say:

* Let's disable `AbcSize`, `CyclomaticComplexity`, and `PerceivedComplexity`. I don't find these to be particular helpful metrics. It is easy enough to judge whether a method is complex or difficult to read.
* As for `MethodLength` and `BlockNesting`, since there are not hundreds of violations I wonder if it would be effective in this case to use `# rubocop:disable`. But first, I'd like to know where these are happening?

mcmire commented on .rubocop.yml:
```diff
@@ -137,6 +147,8 @@ Style/EvenOdd:
   Enabled: false
 Style/FormatString:
   Enabled: false
+Style/FormatStringToken:
```
Did the upgrade force this, or is this an addition?

mcmire replied:
Oh ok gotcha. 👍

mcmire commented:
@utkarsh2102 Thank you for the suggestions. I did notice a lot of changes around quotes and I agree that ideally double quotes make more sense and lead to more consistency. I'm not sure what the ratio of single quote to double quote usage is, though, so it might be a wash. Also, I would rather keep this PR focused to fixing existing violations as it relates to the current state of `.rubocop.yml` (or at least a version of `.rubocop.yml` that works with 0.80.0`) rather than introduce new cops.

KapilSachdev commented:
Thanks for reviewing. I will check those one by one. (Also I just noticed your comment that was made 4 days ago which i missed somehow)

Below i'm responding against the comments made in general.

>It's better to use double_quotes and also this codebase uses it anyway!

The code base has enforced style of `single_quotes`.

>I'm not sure what the ratio of single quote to double quote usage is, though, so it might be a wash.

In master there are `4195` offences if we go towards `double_quotes`, as we have used single quotes in all our specs. and `438` for `single_quotes` which this PR has made changes against.

>Can we stick to an 80-character line length?

Sure, I changed it to 120 thinking that `rubocop` had `80` before and we might have been using 80 because of that and now `rubocop` has changed to use `120` so i updated against that. If you want it to be 80, I will change it back, should I?

>When it comes to indentation, in general I (and to my knowledge this was a rule at thoughtbot, so it doesn't come from nowhere) try not to add more spaces than is necessary

I will look what could be done in that regards as indentation sometimes conflicts with other cops

>There seem to be a lot of changes around trailing commas. 
I too didn't liked it but the there was enforced style for it already, so i thought this might have been discussed before and had been followed. should I revert that as well?

> There are a few cases where a long string was reformatted to fit across multiple lines but now contain line breaks that weren't there before, so make sure to change those back.
There are a few false negatives that I ran across, so make sure to look out for those and change those back too.

Sure, will do so.
I  should not have marked the last commit as [ci skip] otherwise we could have known the test report as well. Though i ran tests against non-cosmetic changes, but not all as `bundle exec appraisal rails_6_0 rpsec` was failing but i didn't debugged further and executed `single` test as they are running fine.

>Let's disable AbcSize, CyclomaticComplexity, and PerceivedComplexity
:+1: 

#### **All 4 `BlockNesting` offenses are under this block.** 
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_model/allow_value_matcher.rb#L432

#### **`Metrics/MethodLength`: 7 offenses**
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_model/allow_value_matcher.rb#L432
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_record/define_enum_for_matcher.rb#L255

Below one can be converted into a constant?
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/spec/unit/shoulda/matchers/active_model/validate_numericality_of_matcher_spec.rb#L5

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/spec/unit/shoulda/matchers/active_model/helpers_spec.rb#L104

Add a `Exclude` for tasks directory?
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/tasks/documentation.rb#L9

The below two locations can be fixed by just removing empty line!
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_record/validate_uniqueness_of_matcher.rb#L932
https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/spec/unit/shoulda/matchers/active_record/validate_uniqueness_of_matcher_spec.rb#L1574

utkarsh2102 commented:
Hello @KapilSachdev 👋🏻 

> The code base has enforced style of single_quotes.
> In master there are 4195 offences if we go towards double_quotes, as we have used single quotes in all our specs. and 438 for single_quotes which this PR has made changes against.

Eeks, I am sorry for not looking more closely! In that case, this looks good, thanks for your work! 💪🏻 

---

Hello @mcmire 👋🏻 

> I did notice a lot of changes around quotes and I agree that ideally double quotes make more sense and lead to more consistency. I'm not sure what the ratio of single quote to double quote usage is, though, so it might be a wash.

Yep, why I say `double_quotes` are better because in cases of string interpolation, you have to use double quotes anyway, there's no escape from that. And if you use single quotes otherwise, it causes inconsistencies (`double quotes` for interpolation and `single_quotes` for other things!). So the best way out is using `double_quotes` bu default!

About the ratio, here's a good research study: https://github.com/rubocop-hq/rubocop/issues/5306#issuecomment-354449447
Over a span of 2.5 years, more people have migrated to using `double_quotes` because of the inconsistency (as mentioned above!).
Furthermore, `bundler` now (already!) ships a default `.rubocop.yml` so the gems by default will now use `double_quotes`. PR here: https://github.com/rubygems/rubygems/pull/3731

Let me know if you think it's worth migrating to `double_quotes` here as well?
If so, I'll be happy to take this after this PR is merged! 😄 

> Also, I would rather keep this PR focused to fixing existing violations as it relates to the current state of .rubocop.yml (or at least a version of .rubocop.yml that works with 0.80.0`) rather than introduce new cops.

You're right! Let's tackle this later as a different PR (as proposed above^)! 👍🏻

KapilSachdev commented on gemfiles/rails_4_2.gemfile:
```diff
@@ -1,36 +1,36 @@
 # This file was generated by Appraisal
 
-source "https://rubygems.org"
+source 'https://rubygems.org'
```
In that case I will add `Bundler/OrderedGems` in the rubocop yml to exclude the `gemfiles` or we not interested in that cop altogether?

mcmire replied:
I think that's fine to have that cop. I think we should exclude Rubocop from parsing `*.gemfiles` altogether, though, since they will never be modified on their own.

KapilSachdev commented on .rubocop.yml:
```diff
@@ -15,23 +15,24 @@ Layout/EmptyLineBetweenDefs:
 Layout/HeredocIndentation:
   Enabled: false
 Layout/LineLength:
+  Exclude:
```
Sure, will add `- !ruby/regexp /\A +(it|describe|context|shared_examples|include_examples|it_behaves_like) ["']/` for all, or the older one is preferred ?

KapilSachdev commented on .rubocop.yml:
```diff
@@ -15,23 +15,24 @@ Layout/EmptyLineBetweenDefs:
+    - '^[ ]*#.+$'
+    - '^[ ]*''.+?'' => ''.+?'',?$'
+    - '^[ ]*".+?" => ".+?",?$'
+  Max: 120
```
Sure, we can keep 80, but i changed to 120 thinking that 80 was because of rubocop's default that time, which is not changed to 120. So should we stay on 80 then?

mcmire replied:
Yeah, I think it makes sense to stay on 80.

KapilSachdev commented on custom_plan.rb:
```diff
@@ -41,12 +41,11 @@ def best_available
     end
 
     require_relative 'spec/support/unit/load_environment'
-  rescue Gem::LoadError => error
```
Disabling...

KapilSachdev commented on custom_plan.rb:
```diff
@@ -41,12 +41,11 @@ def best_available
 
-  def after_fork
-  end
+  def after_fork; end
```
Should we enforce [Style/EmptyMethod](https://docs.rubocop.org/rubocop/cops_style.html#styleemptymethod) with `expanded` or just disable the cop, as i encountered few places where we are using single line empty methods.

mcmire replied:
Oh really? I didn't realize. Okay in that case I think enforcing `expanded` makes more sense.

KapilSachdev commented on lib/shoulda/matchers/action_controller/permit_matcher.rb:
```diff
@@ -303,11 +303,11 @@ def format_parameter_names(parameter_names)
-            else
-              options = {}
-            end
+            options = if subparameter_name
```
WIll check that, as even the default one which is new_line should had compliant about [Layout/MultilineAssignmentLayout](https://docs.rubocop.org/rubocop/cops_layout.html#layoutmultilineassignmentlayout)

KapilSachdev commented on lib/shoulda/matchers/action_controller/render_with_layout_matcher.rb:
```diff
@@ -64,11 +64,11 @@ def render_with_layout(expected_layout = nil)
-          else
-            @expected_layout = nil
-          end
+          @expected_layout = if expected_layout
```
Yes it can enforce it and even its the default [Layout/MultilineAssignmentLayout](https://docs.rubocop.org/rubocop/cops_layout.html#layoutmultilineassignmentlayout). I;m not sure why it didn;t compliant, might be a bug ? Will see to it and will report an issue/PR in rubocop if it is.

KapilSachdev commented on lib/shoulda/matchers/active_model/numericality_matchers/comparison_matcher.rb:
```diff
@@ -76,7 +77,7 @@ def all_bounds_correct?
 
           def failing_submatchers
             submatchers_and_results.
-              select { |x| !x[:matched] }.
```
Sure, will disable [Style/InverseMethods](https://docs.rubocop.org/rubocop/cops_style.html#styleinversemethods)

KapilSachdev commented on lib/shoulda/matchers/active_record/validate_uniqueness_of_matcher.rb:
```diff
@@ -528,20 +528,15 @@ def find_or_create_existing_record
-          else
-            nil
-          end
+          record.presence
```
Oh yes! what was i thinking ...

KapilSachdev commented on lib/shoulda/matchers/doublespeak/double_implementation_registry.rb:
```diff
@@ -4,7 +4,9 @@ module Doublespeak
       # @private
       module DoubleImplementationRegistry
         class << self
+          # rubocop:disable Style/MutableConstant
```
That would be good.

KapilSachdev commented on lib/shoulda/matchers/integrations/libraries/action_controller.rb:
```diff
@@ -13,9 +13,8 @@ def integrate_with(test_framework)
             test_framework.include(matchers_module, type: :controller)
 
             include_into(::ActionController::TestCase, matchers_module) do
-              def subject
```
[Lint/NestedMethodDefinition](https://docs.rubocop.org/rubocop/cops_lint.html#lintnestedmethoddefinition)

KapilSachdev commented on lib/shoulda/matchers/util/word_wrap.rb:
```diff
@@ -169,7 +169,7 @@ def read_indentation
       end
 
-      def wrap_line(line, direction: :left)
+      def wrap_line(line, _direction: :left)
```
I think its safe to remove remove as the method definition is not caring for this.

KapilSachdev commented on spec/support/tests/filesystem.rb:
```diff
@@ -63,9 +63,9 @@ def create_parents_of(path)
-    def append_to_file(path, content, options = {})
+    def append_to_file(path, content, _options = {})
       create_parents_of(path)
-      open(path, 'a') { |f| f.puts(content + "\n") }
```
The complain was from [Security/Open](https://docs.rubocop.org/rubocop/cops_security.html#securityopen). I see that there is an open method, maybe rubocop is to blame here. Will check with rubocop and fix there in case found guilty!

KapilSachdev commented on spec/support/unit/helpers/confirmation_matcher_helpers.rb:
```diff
@@ -10,8 +10,7 @@ def builder_for_record_validating_confirmation(options = {})
     def builder_for_record_validating_confirmation_with_18n_message(options = {})
       builder = builder_for_record_validating_confirmation(options)
       RecordBuilderWithI18nValidationMessage.new(builder,
```
Will verify why!

KapilSachdev commented on spec/unit/shoulda/matchers/active_model/allow_value_matcher_spec.rb:
```diff
@@ -434,26 +434,26 @@
-        expect(validating_format(with: /abc/, on: :customisable)).to allow_value("abcde").for(:attr).on(:customisable)
+    context 'with the validation context' do
+      it 'allows a good value' do
+        expect(validating_format(with: /abc/, on: :customisable)).to allow_value('abcde').for(:attr).on(:customisable)
```
Sure, please confirm if we are going down to 80.

KapilSachdev commented on spec/unit/shoulda/matchers/active_model/allow_value_matcher_spec.rb:
```diff
@@ -464,7 +464,7 @@
         validates_presence_of     :attr
         validates_length_of       :attr, within: 1..5
         validates_numericality_of :attr, greater_than_or_equal_to: 1,
```
I will try to restore this change, but i think adding parenthesis doesn't go with the consistency! 
Will use this in case didn't get around.

KapilSachdev commented on spec/unit/shoulda/matchers/active_model/disallow_value_matcher_spec.rb:
```diff
@@ -69,12 +68,8 @@
-            self.errors[:attr2] << 'some message'
-          end
-        end
+        custom_validation = -> { errors[:attr2] << 'some message' if self[:attr] != 'good value' }
```
Lint/NestedMethodDefinition

mcmire commented:
@utkarsh2102 Hi just wanted to respond to your comment. I realize that most people probably use double quotes now but I'd rather not change the entire codebase to use double quotes. Although it may seem like there are bunch of changes to the code style in this PR, the code is really being aligned to fit rules that have already been in place (or enforce unofficial rules). It was bound to happen anyway so I treat this as a rip-the-bandaid-off kind of situation. But I'd rather not change the existing style at this point and force a widespread code change. I appreciate your suggestion though :)

utkarsh2102 commented:
Hi @mcmire,

No problem!
Thanks for all your work :)

mcmire commented:
@KapilSachdev Sorry for the delay in responding to your last comment. Here are my responses for the warnings you mentioned, first giving the link to the code, then the commentary:

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_model/allow_value_matcher.rb#L432

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_record/define_enum_for_matcher.rb#L255

The warnings for these methods make sense — they are super long. I think refactoring them is out of scope, but it's probably worth noting. Maybe for now we want to disable BlockNesting and MethodLength for them and make a new issue to refactor them later?

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/spec/unit/shoulda/matchers/active_model/validate_numericality_of_matcher_spec.rb#L5

This one seems okay to me. There's no real way to make a constant inside of a block, because that constant will get hoisted outside that block, so I think our only choice is to have a method. I think disabling MethodLength is fine for this.

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/spec/unit/shoulda/matchers/active_model/helpers_spec.rb#L104

I feel like there might be a way to refactor this slightly, but even so it will likely be over the limit still. Perhaps we disable MethodLength for this one as well.

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/tasks/documentation.rb#L9

Maybe we ought to refactor this into different methods: `add_generate_task`, `add_autogenerate_task`, etc. That would probably be more bite-size. If you want to do this now that's fine, otherwise I would disable MethodLength for this method. (Again I'd rather avoid `Exclude`ing an entire directory unless we know there are 1000 violations or something.)

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/lib/shoulda/matchers/active_record/validate_uniqueness_of_matcher.rb#L932

This method is super long as well, so ideally it could be refactored later, but if removing the empty lines at the end keep it under the limit I'm fine with that.

https://github.com/thoughtbot/shoulda-matchers/blob/7cedd3e4e199eaf589e56457f9a1ff97995af938/spec/unit/shoulda/matchers/active_record/validate_uniqueness_of_matcher_spec.rb#L1574

There might be a way to refactor this as well although it's not obvious, so removing the empty lines here makes sense.

KapilSachdev commented:
Thanks for the feedback @mcmire and no worries. I didn't get the time to work on it further and might not get this week also.

> Maybe for now we want to disable BlockNesting and MethodLength for them and make a new issue to refactor them later?

:+1:  I actually wanted to get hound or whatever we choose to get back up and running so that it will help cut your review cost a bit, and meanwhile we can refactor the remaining pieces. 

> Again I'd rather avoid `Excluding` an entire directory unless we know there are 1000 violations or something.

I actually do prefer the code to be consistent with rules defined, around all areas of source code, be it either script/tasks, specs etc... but i think i wanted to get rid of those at that point. Its better we ignore them and add to `TODO` refactoring list.

mcmire commented:
@KapilSachdev Reading back over your comment, I realize this PR is pretty large so I'm fine with Excluding things for now if it helps to get this across the finish line.

KapilSachdev commented:
@mcmire Apologies for late reply. I will try to complete the suggestions you made soon and we can decide on next steps.

mcmire commented:
@KapilSachdev No worries! Sounds good.

KapilSachdev commented:
All Cops are handled, but `Style/StringLiterals` is a problem.
`Appraisal` seems to be using `double_quotes` when generating `gemfiles`, which always creates unnecessary offences with our current `single_quotes` enforcement of concerned cop.

mcmire commented:
@KapilSachdev Gotcha. Yeah, since nothing in `gemfiles` is meant to be editable directly, I think this would be a case where I would be totally fine with Excluding everything there.

KapilSachdev commented:
Excluding/Ignoring this [`Style/StringLiterals`](https://docs.rubocop.org/rubocop/cops_style.html#stylestringliterals) cop is not supported.

mcmire commented:
@KapilSachdev That seems odd, but okay. I mean, the good news is that I don't expect the Appraisal gemfiles to be updated much, so if we get some violations there I don't think it's the end of the world.

Given that, I'm good with this PR if you are? I don't know if you want to squash this branch or leave everything separate (I'm fine with either).

VSPPedro commented:
What do you think about updating Rubocop to the most current version? (0.93.1)

(In case it is necessary to give only one bundle update.)

KapilSachdev commented:
@mcmire Sure we can leave it as is.
I'm good with the current changes. And as for squashing, If you want, i will do that.

@VSPPedro All rubocop gems are upto date already.

VSPPedro commented:
I'm so sorry. I saw the commit `Bump rubocop from 0.89.1 to 0.90` and thought that the gem was not with the latest version. I should have looked at Gemfile.lock. My bad.

mcmire commented:
Just checked CI and it looks like there are a couple of failures. Perhaps one of the `.freeze` changes caused this? Kind of a weird error.

KapilSachdev commented:
@VSPPedro I should have added this in the commit message. This can was not done in the commit you are mentioning, but rather in the changes pushed yesterday.

@mcmire Yes, it is weird. The failing test should have failed in all the test suites, but they didn't.

```ruby
      class ValidateNumericalityOfMatcher
        NUMERIC_NAME = 'number'.freeze
        NON_NUMERIC_VALUE = 'abcd'.freeze # This seems to be the issue
        DEFAULT_DIFF_TO_COMPARE = 1
...
```
I will look into it.
Do you have any thoughts on it as to why this is happening?

mcmire commented:
@KapilSachdev It looks like the test that's failing involves setting a value on an attribute which is a Postgres `money` column type. It looks like the `money` type, in the process of type-casting the given value, directly modifies the incoming value using `sub!`. Since the incoming value is frozen in this case, that fails. This feels like a defect in Rails, but since we can't do anything about this, perhaps the fix here is to move [`NON_NUMERIC_VALUE` in validate_numericality_of_matcher.rb](https://github.com/thoughtbot/shoulda-matchers/pull/1350/files#diff-8328f6df41a2e997135b84d7555df2ee17ddb185865b1787eec1f51ca6e4a539L533) so that it's a class or instance method instead of a constant, and then take off `.freeze`?

KapilSachdev commented:
@mcmire As `gemfiles` are auto-generated I have excluded them  from `rubocop`.
All green now

KapilSachdev commented:
Upgrading to v1.0 in current PR state will add the following offences.

| Count | Cop                                      |
|-------|------------------------------------------|
| 54    | Style/StringConcatenation                |
| 10    | Style/OptionalBooleanParameter           |
| 4     | Layout/EmptyLinesAroundAttributeAccessor |
| 3     | Style/HashLikeCase                       |
| 2     | Lint/MissingSuper                        |
| 2     | Style/RedundantRegexpCharacterClass      |
| 1     | Layout/SpaceAroundMethodCallOperator     |
| 1     | Lint/EmptyExpression                     |
| 1     | Lint/EmptyInterpolation                  |
| 1     | Style/CaseLikeIf                         |
| 1     | Style/ClassEqualityComparison            |
| 1     | Style/ExplicitBlockArgument              |
| 1     | Style/GlobalStdStream                    |

VSPPedro commented:
IMO, I think it would be best to let this upgrade for another PR.

KapilSachdev commented:
Yes, I'm working on it. Just mentioned the cops in case we want any cop to be disabled or need some non-default enforcement preference.

KapilSachdev commented:
@mcmire Please confirm the expectation for `Lint/MissingSuper`. Do we want to disable it, or ignore for the instances mentioned below 

https://github.com/thoughtbot/shoulda-matchers/blob/ffd5a28b451f3589408c6893908c6c3956c53801/lib/shoulda/matchers/doublespeak/object_double.rb#L21-L26

For `Style/HashLikeCase` following changes were done
https://github.com/thoughtbot/shoulda-matchers/blob/ffd5a28b451f3589408c6893908c6c3956c53801/lib/shoulda/matchers/active_model/numericality_matchers/comparison_matcher.rb#L6-L14
```ruby
ERROR_MESSAGES = {
            :> => { label: :greater_than,
                    assertions: [false, false, true], },
            :>= => { label: :greater_than_or_equal_to,
                     assertions: [false, true, true], },
            :< => { label: :less_than,
                    assertions: [true, false, false], },
            :<= => { label: :less_than_or_equal_to,
                     assertions: [true, true, false], },
            :== => { label: :equal_to,
                     assertions: [false, true, false], },
            :!= => { label: :other_than,
                     assertions: [true, false, true], },
          }.freeze
```

and 

https://github.com/thoughtbot/shoulda-matchers/blob/ffd5a28b451f3589408c6893908c6c3956c53801/lib/shoulda/matchers/active_record/association_matcher.rb#L1188-L1199
```ruby
        MACROS = {
          'belongs_to' => 'belong to',
          'has_many' => 'have many',
          'has_one' => 'have one',
          'has_and_belongs_to_many' => 'have and belong to many',
        }.freeze
```

mcmire commented on lib/shoulda/matchers/active_model/numericality_matchers/comparison_matcher.rb:
```diff
@@ -5,12 +5,18 @@ module NumericalityMatchers
-            :<= => :less_than_or_equal_to,
-            :== => :equal_to,
-            :!= => :other_than,
+            :> => { label: :greater_than,
```
Can you format this as the following:

``` ruby
{
  label: :greater_than,
  assertions: [false, false, true]
},
...
```

Again, abiding by the ~rule~ guideline that you shouldn't need to add extra spaces to align items.

mcmire commented on lib/shoulda/matchers/active_model/numericality_matchers/comparison_matcher.rb:
```diff
@@ -86,19 +92,20 @@ def last_failing_submatcher
-                matcher.with_message(@message, values: { count: @value })
-                matcher
-              end
+            @_submatchers ||= comparison_combos.map do |
```
This seems like another change to please Rubocop but actually detracts from the readability, would you mind changing this back?

mcmire replied:
Haha, no worries!

mcmire commented on lib/shoulda/matchers/active_model/numericality_matchers/comparison_matcher.rb:
```diff
@@ -77,7 +83,7 @@ def all_bounds_correct?
           def failing_submatchers
             submatchers_and_results.
-              reject { |x| x[:matched] }.
+              select { |x| !x[:matched] }.
```
I know I've asked about this before but which cop controls this? This seems like an unnecessary check.

KapilSachdev replied:
This was reverted to its original state.
`Style/InverseMethods` controls this behaviour.

mcmire replied:
Yeah, sorry for the bluntness. I just think it may lead to churn.

mcmire commented on lib/shoulda/matchers/active_model/validate_numericality_of_matcher.rb:
```diff
@@ -558,7 +558,8 @@ def add_submatcher(submatcher)
 
           if submatcher.respond_to?(:diff_to_compare)
-            @diff_to_compare = [@diff_to_compare, submatcher.diff_to_compare].max
+            @diff_to_compare = [@diff_to_compare,
```
Maybe:

``` ruby
@diff_to_compare = [
  @diff_to_compare,
  submatcher.diff_to_compare,
].max
```

Again, the whole "keep indentation to multiple of 2 and don't add more than is necessary" sort of thing

mcmire commented on lib/shoulda/matchers/active_record.rb:
```diff
@@ -1,6 +1,7 @@
 require 'shoulda/matchers/active_record/association_matcher'
 require 'shoulda/matchers/active_record/association_matchers'
-require 'shoulda/matchers/active_record/association_matchers/counter_cache_matcher'
+require 'shoulda/matchers/active_record/association_matchers/'\
```
Another thing I don't think is necessary :)

mcmire commented on lib/shoulda/matchers/active_record/association_matcher.rb:
```diff
@@ -990,6 +990,13 @@ class AssociationMatcher
 
         attr_reader :name, :options
 
+        MACROS = {
```
Would you mind putting this at the top of the class? I believe that we have been pretty good about putting them first.

mcmire commented on lib/shoulda/matchers/active_record/association_matcher.rb:
```diff
@@ -1292,7 +1299,8 @@ def class_name_correct?
 
         def join_table_correct?
-          if macro != :has_and_belongs_to_many || join_table_matcher.matches?(@subject)
+          if macro != :has_and_belongs_to_many || \
```
When it comes to stretching if statements across multiple lines, what do you think about formatting this as:

``` ruby
if (
  macro != :has_and_belongs_to_many ||
  join_table_matcher.matches?(@subject)
)
```

mcmire commented on lib/shoulda/matchers/active_record/association_matchers/required_matcher.rb:
```diff
@@ -33,9 +33,9 @@ def matches?(subject)
                 end
 
               missing_option << (
-                'fail validation if ' +
```
Is this enforced by a cop? I notice you doing this a lot and I'm wondering if it's your style or something that Rubocop is enforcing.

KapilSachdev replied:
This change is enforced by `1.0`, which requires string concatenation to be enforced with interpolation by default.
So to counter LineLength and [Style/StringConcatenation](https://docs.rubocop.org/rubocop/cops_style.html#stylestringconcatenation) cops, i had to use `\`.
I was thinking of `heredoc`, but that would be too much of a change, so updated it with `\` for your feedback.

mcmire replied:
Okay, gotcha 👍

mcmire reviewed (commented):
Adding some more comments. We're getting close. One thing I would say is that I feel like in some cases we are making code changes just to appease Rubocop but they actually detract from the readability or make things look objectively awkward. For instance, needing to add `\\` to appease the line length. For now it's _probably_ okay just so we can get this over the finish line, but maybe it's something we should look to relax in the future.

mcmire commented:
@KapilSachdev Let's leave both `Style/OptionalBooleanParameter` and `Lint/MissingSuper` disabled for now.

VSPPedro commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -67,7 +69,7 @@ def has_attribute?
 
         def has_expected_action_text?
-          @subject.send(rich_text_attribute).class.name == 'ActionText::RichText'
+          @subject.send(rich_text_attribute).instance_of?(ActionText::RichText)
```
Instead of using the class, can you try using a string?

```ruby
@subject.send(rich_text_attribute).instance_of?('ActionText::RichText')
```

Let's see if this change will fix the problem on CI. I almost sure. I'll make some tests.

KapilSachdev replied:
class name evaluation with string poses an issue of being failed silently when the class is not even defined.
But this current scenario is not one of those cases so we may revert to string class comparison.

VSPPedro replied:
I did not know that. Thank you!

KapilSachdev commented on lib/shoulda/matchers/active_model.rb:
```diff
@@ -4,12 +4,16 @@
 require 'shoulda/matchers/active_model/allow_value_matcher'
-require 'shoulda/matchers/active_model/allow_value_matcher/attribute_changed_value_error'
-require 'shoulda/matchers/active_model/allow_value_matcher/attribute_does_not_exist_error'
+require 'shoulda/matchers/active_model/allow_value_matcher/'\
```
I will check if `LineLength` can be disable for require. It should work.

mcmire replied:
Okay perfect :)

mcmire commented:
@KapilSachdev Anything else you noticed? I'm good with this PR if you are.

KapilSachdev commented:
@mcmire I think the current state of the PR is fine.

mcmire commented:
@KapilSachdev Okay cool. I see you keep force-pushing, so did you want to keep your commits separate or put them together?

KapilSachdev commented:
@mcmire  I had wanted to keep them separate, but if you think the change should go in one commit altogether, thats fine with me too.
Should I squash them all in one?

mcmire commented:
@KapilSachdev No, it's fine with me to keep them separate! Just wanted to know your preference.

Merging now!

KapilSachdev commented:
Thank you.
